README.md
# blockify

Blockify is a linux only application that allows you to automatically mute songs and advertisements in Spotify.

**Important**: If you are using the 'old' Spotify client (version < 1.0), please use the legacy-branch (== blockify version < 2.0).


## Installation

### Dependencies

Mandatory:
  - Python3
  - Spotify > 1.0.12 (to get the latest version follow the instructions given here [Spotify-Client-1-x-beta-] (https://community.spotify.com/t5/Spotify-Community-Blog/Spotify-Client-1-x-beta-for-Linux-has-been-released/ba-p/1147084))
  - wmtrcl (provides information about the Spotify window)

Optional but highly recommended:
  - Pulseaudio (allows muting Spotify instead of all system sound)
  - Gstreamer1.0 (used to play music of your choice during muted ads. Requires pulseaudio.)

Optional:
  - docopt (provides a command-line interface for blockify and blockify-ui)

On ArchLinux, you can install all dependencies as follows:  
`pacman -S git python-pip gst-python pulseaudio alsa-utils pygtk python-dbus python-gobject python-docopt wmctrl`

### Package Manager

#### Arch Linux

Available via AUR here: https://aur.archlinux.org/packages/?O=0&K=blockify

Example ArchLinux installation routine:  
``` bash
git clone https://aur.archlinux.org/blockify.git
cd blockify
makepkg -sri
```

#### Fedora and openSUSE

Available in the [openSUSE build service](https://build.opensuse.org/package/show/home:fusion809/blockify).

### Manual (pip/setup.py)

If there is no blockify package available on your distribution, you'll have to install it directly via one of pythons many installation tools.  


Package names are for ArchLinux and will probably differ slightly between distributions.

Preparation (for Ubuntu):  
```bash
# Install Spotify beta
echo deb http://repository.spotify.com testing non-free | sudo tee /etc/apt/sources.list.d/spotify.list
sudo apt-get update
sudo apt-get install spotify-client
# Install blockify dependencies
sudo apt-get install git python3-pip python3-gst-1.0 python3-requests python3-docopt wmctrl
```

Install routine:  
```bash
# Install blockify
git clone https://github.com/serialoverflow/blockify
cd blockify
sudo pip3 install .
echo -e '[Desktop Entry]\nName=Blockify\nComment=Blocks Spotify commercials\nExec=blockify-ui\nIcon='$(python3 -c 'import pkg_resources; print(pkg_resources.resource_filename("blockify", "data/icon-red-512.png"))')'\nType=Application\nCategories=AudioVideo' | sudo tee /usr/share/applications/blockify.desktop
```

## Usage

### Requirements

It is important to know that blockify relies on dbus (and, for some features, on pulseaudio) for ad detection.  
If any of these statements are true for your configuration, ad detection will _not_ work:  
* DBus is disabled
* Spotify is minimized to the system tray (task bar is fine)
* Notifications are disabled in Spotify  

Additionally, blockify makes use of pulseaudio sinks, allowing processes to be muted individually.    
If you do not have/want pulseaudio, blockify will mute the system sound during commercials instead of just Spotify. The interlude music feature will not work as a consequence.

### Detection

Blockify will automatically detect and block ads for you so besides starting it after running spotify, there's not a lot to do.  
However, it also comes with the option to complement or replace that autoblock functionality with a blocklist (saved as ~/.config/blockify/blocklist.txt).  
Blocklist entries are case-sensitive and greedy, e.g. the entry `Blood` would match any artist starting with those exact five letters.    

### Controls/Actions

Blockify accepts several signals:
* SIGINT(9)/SIGTERM(15): Exit cleanly.
* SIGUSR1(10): Block current song.
* SIGUSR2(12): Unblock current song.
* SIGRTMIN(34): Play previous spotify song.
* SIGRTMIN+1(35): Play next spotify song.
* SIGRTMIN+2(36): Toggle play/pause the current spotify song.
* SIGRTMIN+3(37): Toggle block state of current song.
* SIGRTMIN+10(44): Play previous interlude song.
* SIGRTMIN+11(45): Play next interlude song.
* SIGRTMIN+12(46): Toggle play/pause the current interlude song.
* SIGRTMIN+13(47): Toggle interlude autoresume.

To easily use these signals add the following function to your .bashrc:
```bash
bb() {
    local signal
    local cmd
    [[ "$#" -lt 1 ]] && echo "Usage: bb ( b[lock] | u[nblock] | p[revious] | n[ext] | t[oggle] | t[oggle]b[lock] |...)"  && return 0
    case "$1" in
        '')  blockify-dbus get 2>/dev/null && return 0;;
        ex|exit)
            signal='TERM';;       # Exit
        b|block)
            signal='USR1';;       # Block
        u|unblock)
            signal='USR2';;       # Unblock
        p|previous)
            signal='RTMIN';;      # Previous song
        n|next)
            signal='RTMIN+1';;    # Next song
        t|toggle)
            signal='RTMIN+2';;    # Toggle play song
        tb|toggleblock)
            signal='RTMIN+3';;    # Toggle block song
        ip|iprevious)
            signal='RTMIN+10';;   # Previous interlude song
        in|inext)
            signal='RTMIN+11';;   # Next interlude song
        it|itoggle)
            signal='RTMIN+12';;   # Toggle play interlude song
        itr|itoggleresume)
            signal='RTMIN+13';;   # Toggle interlude resume
        *) echo "Bad option" && return 0;;
    esac
    pkill --signal "$signal" -f 'python.*blockify'
}
```
Then use it via e.g. `bb` to get current song info or `bb t` to toggle playback.

#### CLI

Blockify has a CLI/daemon that you can start with `blockify`.  
`blockify -h` will print out a help text with available options.  

##### GUI

Alternatively, you can use the GUI with `blockify-ui` which spawns this window.  
![ScreenShot](gui_screenshot.png)
- Play, Previous, Next: These buttons use dbus to send audio control commands to spotify.
- Block/Unblock: Add/Remove the currently playing song to/from the blocklist.
- Mute/Unmute: Mute/Unmute the current song. Only works if "Manual" checkbox is activated.
- Mute (manual): Disables automatic mute of ads and instead allows you to mute manually.
- Show/Hide Cover: Enable/Disable display of cover art image.
- Autodetect: Let blockify automatically detect ads or switch to manual blocklist-style.
- Automute: Let blockify automatically handle mute or switch to manual muting.
- Autohide: If this option is checked, the cover art will be automatically hidden whenever a commercial is playing.
- Open/Close List: Opens a small popup text editor with the blocklist opened where you can edit and save it. Keybinds: Control-S (save), Control-W/Q (close), Control-D (delete current line).
- Exit: Stop blockify cleanly, i.e. unmute sinks, update blocklist and generally clean up.
- Enable/disable Player: Shows and enables resp. hides and disables the mini audio player (interlude player) below the button which will play music instead of commercials.
- Prev, Play/Pause, Next (bottom): Media buttons to control the interlude player.
- Browse: Allows you to open m3u-playlists and/or audio files on the fly. You can select multiple items and combine playlists with audio files. The player will automatically load those and discard the previous playlist.  
- Autoresume: If enabled, the interlude player will switch back to Spotify as soon as the commercials end. If disabled, the current interlude song will be finished before switching back to Spotify.  

#### Configuration

Please see the provided [example_blockify.ini](https://github.com/serialoverflow/blockify/blob/master/blockify/data/example_blockify.ini) on what settings are available and their purpose.  
Blockify automatically creates a configuration file at `$HOME/.config/blockify/blockify.ini` if you don't have one already. It will also tell you via ERROR-logging messages, if you configuration file is faulty or incomplete, in which case the options that could be read will be merged with the default options you see in example_blockify.ini but you'll still want to fix your configuration file.  

#### Interlude Music

From version 1.4 onwards blockify can play music of your choice during commercial breaks.  
The default behaviour is for blockify to automatically play the first song in the playlist file (should you have one), when a commercial starts playing.  
Alternatively, you can set the autoresume option to False which will cause blockify to always finish the current interlude song before resuming spotify playback.   

The interlude feature only works if you use pulseaudio (i.e. spotify needs to run in its own sink so it can be muted separately).    
To make use of interlude music you have to configure a playlist file in `~/.config/blockify/playlist.m3u`.  
The playlist system is (mostly) M3U-compliant.  

An example playlist:
```
# Lines starting with "#" will be ignored.
# Absolute path to a file:
/media/music/foo/bar.mp3
# Relative path to a file (as seen from playlist location):
foo/bar.flac
# Relative path to another playlist. Just make sure the other playlist doesn't link back or else you'll get a very long playlist:
baz.m3u
# A whole directory:
/media/music/foo
# It's also possible to give full URIs:
file:///media/music/foo/bar.mp4
http://www.example.com/foo/bar.wav
# A radio station. Note that radio streams don't usually end so you'll have to switch
# back to spotify manually, enable autoresume or specify the radio_timeout in the config file.
http://skyserver5.skydisc.net:8000
```

CHANGELOG.md
# Changelog

- v3.6.3 (2016-05-20): Fix [issue #105](https://github.com/serialoverflow/blockify/issues/105) by adding new general option "use_window_title".
- v3.6.2 (2016-05-08): Fix issue where pulse sink would sometimes not be unmuted on exit, fix false positive ad detection on pause for people not using pulseaudio sinks ([issue #97](https://github.com/serialoverflow/blockify/issues/97)), fix pypi installation ([issue #99](https://github.com/serialoverflow/blockify/issues/99)), include README.rst, remove requirements.txt and reimplement dbusclient.py CLI.
- v3.6.1 (2016-04-11): Fix Gst initialization in interlude player, refactor dbusclient CLI and improve some documentation.
- v3.6.0 (2016-04-10): Published blockify to PyPI, made it virtualenv compatible (still requires --system-site-packages for gi.repository) and refactored the import structure.
- v3.5.0 (2016-04-08): Reintroduce wmctrl to catch video ads ([issue #89](https://github.com/serialoverflow/blockify/issues/89)) and block some audio ads more reliably. Fix encoding issues ([issue #95](https://github.com/serialoverflow/blockify/issues/95)).
- v3.4.0 (2016-03-25): Fix play/pause toggle button, right click on tray [issue #83](https://github.com/serialoverflow/blockify/issues/83) and add start_minimized option [issue #93](https://github.com/serialoverflow/blockify/issues/93).
- v3.3.1 (2016-01-03): Fix interlude player crashes ([issue #84](https://github.com/serialoverflow/blockify/issues/84)).
- v3.3.0 (2016-01-03): Enable profiling, improve GUI performance, fix playback button & title status functionality and add tray icon toolip.
- v3.2.1 (2016-01-03): Remove unnecessary imports and other cleanups.
- v3.2.0 (2015-12-31): Reintroduce playback status (see [issue #68](https://github.com/serialoverflow/blockify/issues/68))
- v3.1.0 (2015-12-31): Remove wmctrl dependency (see [issue #67](https://github.com/serialoverflow/blockify/issues/67))
- v3.0.0 (2015-10-16): Remove beta status and port to python3 and gstreamer1.0 (see [issue #59](https://github.com/serialoverflow/blockify/issues/59)).
- v2.0.1 (2015-10-05): (prerelease) Fix [issue #58](https://github.com/serialoverflow/blockify/issues/58) and [issue #38](https://github.com/serialoverflow/blockify/issues/38).
- v2.0.0 (2015-09-05): (prerelease) Added rudimentary support for Spotify v1.0 and higher. Fixed autoplay option.
- v1.9.0 (2015-08-15): Fix [issue #52](https://github.com/serialoverflow/blockify/issues/52), introduce autoplay option and change start_spotify option to boolean type
- v1.8.8 (2015-07-11): Fix [issue #46](https://github.com/serialoverflow/blockify/issues/46) and [issue #47](https://github.com/serialoverflow/blockify/issues/47)
- v1.8.7 (2015-06-11): Pressing play will now properly pause interlude music before resuming spotify playback.
- v1.8.6 (2015-05-10): Minor refactoring and removed incomplete "fix" for [issue #44](https://github.com/serialoverflow/blockify/issues/44).
- v1.8.5 (2015-05-09): Signal cleanups and [issue #44](https://github.com/serialoverflow/blockify/issues/44) again.
- v1.8.4 (2015-05-08): Add additional signals for both spotify and interlude controls (prev/next/playpause, ...), see Controls/Actions section in this README
- v1.8.3 (2015-05-06): Fix [issue #44](https://github.com/serialoverflow/blockify/issues/44): Cancel current interlude song and resume spotify playback if next spotify song button is clicked when no ad is playing
- v1.8.2 (2015-03-18): Reintroduced pacmd_muted_value option in general section ([issue #38](https://github.com/serialoverflow/blockify/issues/38)). Added `gobject.threads_init()` to address ([issue #42](https://github.com/serialoverflow/blockify/issues/42)).
- v1.8.1 (2015-03-17): Added start_shuffled option in interlude-section ([issue #41](https://github.com/serialoverflow/blockify/issues/41))
- v1.8.0 (2015-03-15): Added substring_search option ([issue #36](https://github.com/serialoverflow/blockify/issues/36)). Added pacmd_muted_value option ([issue #38](https://github.com/serialoverflow/blockify/issues/38)). Removed gtk.threads_init() ([issue #39](https://github.com/serialoverflow/blockify/issues/39)).
- v1.7.2 (2015-01-10): Added unmute_delay option for the GUI, too. Removed forced unmute when Spotify is not playing a song or blockify can't find an ad.
- v1.7.1 (2014-12-26): Fix for [issue #32](https://github.com/serialoverflow/blockify/issues/32) (introduced playback_delay option), better load_config and update_slider error catching

blockify/blockify.py
#!/usr/bin/env python3
"""blockify

Usage:
    blockify [-l <path>] [-v...] [-q] [-h]

Options:
    -l, --log=<path>  Enables logging to the logfile/-path specified.
    -q, --quiet       Don't print anything to stdout.
    -v                Verbosity of the logging module, up to -vvv.
    -h, --help        Show this help text.
    --version         Show current version of blockify.
"""
import logging
import os
import re
import signal
import subprocess
import sys
import time
import blockifydbus
import blocklist
import util
from gi import require_version

require_version('Gtk', '3.0')
require_version('Wnck', '3.0')

from gi.repository import Gtk
from gi.repository import Wnck

from gi.repository import GObject

log = logging.getLogger("main")


class Blockify(object):

    def __init__(self, blocklist):
        self.blocklist = blocklist
        self.orglist = blocklist[:]
        self.check_for_blockify_process()
        self.start_spotify_if_necessary()

        self._autodetect = util.CONFIG["general"]["autodetect"]
        self._automute = util.CONFIG["general"]["automute"]
        self.autoplay = util.CONFIG["general"]["autoplay"]
        self.unmute_delay = util.CONFIG["cli"]["unmute_delay"]
        self.update_interval = util.CONFIG["cli"]["update_interval"]
        self.spotify_refresh_interval = 2500
        self.suspend_blockify = False
        self.pulse_unmuted_value = ""
        self.song_delimiter = " - "  # u" \u2013 "
        self.found = False
        self.current_song = ""
        self.previous_song = ""
        self.song_status = ""
        self.is_fully_muted = False
        self.is_sink_muted = False
        self.dbus = self.init_dbus()
        self.channels = self.init_channels()
        # The gst library used by interludeplayer for some reason modifies
        # argv, overwriting some of docopts functionality in the process,
        # so we import gst here, where docopts cannot be broken anymore.
        import interludeplayer
        self.player = interludeplayer.InterludePlayer(self)

        self.initialize_mute_method()

        self.initialize_pulse_unmuted_value()

        # Only use interlude music if we use pulse sinks and the interlude playlist is non-empty.
        self.use_interlude_music = util.CONFIG["interlude"]["use_interlude_music"] and \
                                   self.mutemethod == self.pulsesink_mute and \
                                   self.player.max_index >= 0

        log.info("Blockify initialized.")

    def start_spotify_if_necessary(self):
        if self.check_for_spotify_process():
            return
        log.error("No spotify process found.")

        if not util.CONFIG["general"]["start_spotify"]:
            log.info("Exiting. Bye.")
            sys.exit()

        self.start_spotify()
        if not self.check_for_spotify_process():
            log.error("Failed to start Spotify!")
            log.info("Exiting. Bye.")
            sys.exit()

    def is_localized_pulseaudio(self):
        """Pulseaudio versions below 7.0 are localized."""
        localized = False
        try:
            pulseaudio_version_string = subprocess.check_output("pulseaudio --version | awk '{print $2}'", shell=True)
            pulseaudio_version = int(pulseaudio_version_string[0])
            localized = pulseaudio_version < 7
        except Exception as e:
            log.error("Could not detect pulseaudio version: {}".format(e))

        return localized

    def initialize_pulse_unmuted_value(self):
        """Set 'no' as self.pulse_unmuted_value and try to translate if necessary."""
        unmuted_value = 'no'
        if self.is_localized_pulseaudio():
            try:
                self.install_locale()
                # Translate 'no' to the system locale.
                unmuted_value = _(unmuted_value)
            except (Exception):
                log.debug("Could not install localization. If your system "
                          "language is not english this *might* lead to unexpected "
                          "mute behaviour. A possible fix is to replace the "
                          "value of unmuted_value in blockify.py with your "
                          "translation of 'no', e.g. 'tak' in polish.")

        self.pulse_unmuted_value = unmuted_value

    def initialize_mute_method(self):
        """Determine if we can use sinks or have to use alsa."""
        try:
            devnull = open(os.devnull)
            subprocess.check_output(["pacmd", "list-sink-inputs"], stderr=devnull)
            self.mutemethod = self.pulsesink_mute
            log.debug("Mute method is pulse sink.")
        except (OSError, subprocess.CalledProcessError):
            log.debug("Mute method is alsa or pulse without sinks.")
            log.info("No pulse sinks found, falling back to system mute via alsa/pulse.")
            self.mutemethod = self.alsa_mute

    def install_locale(self):
        import locale
        import gettext

        current_locale, encoding = locale.getdefaultlocale()
        pulseaudio_domain = 'pulseaudio'
        localedir = gettext.find(pulseaudio_domain, languages=[current_locale])
        localedir = localedir[:localedir.find('locale/')] + 'locale'
        locale = gettext.translation(pulseaudio_domain, localedir=localedir, languages=[current_locale])
        locale.install()

    def check_for_blockify_process(self):
        try:
            pid = subprocess.check_output(["pgrep", "-f", "python.*blockify"])
        except subprocess.CalledProcessError:
            # No blockify process found. Great, this is what we want.
            pass
        else:
            if pid.strip().decode("utf-8") != str(os.getpid()):
                log.error("A blockify process is already running. Exiting.")
                sys.exit()

    def check_for_spotify_process(self):
        try:
            pidof_out = subprocess.check_output(["pidof", "spotify"])
            self.spotify_pids = pidof_out.decode("utf-8").strip().split(" ")
            return True
        except subprocess.CalledProcessError:
            return False

    def start_spotify(self):
        if util.CONFIG["general"]["start_spotify"]:
            log.info("Starting Spotify ...")
            subprocess.Popen(['/usr/bin/spotify'])
            for _ in range(20):
                time.sleep(1)
                spotify_is_running = self.check_for_spotify_process()
                if spotify_is_running:
                    log.info("Spotify launched!")
                    break

    def init_channels(self):
        channel_list = ["Master"]
        amixer_output = subprocess.check_output("amixer")
        if "'Speaker',0" in amixer_output.decode("utf-8"):
            channel_list.append("Speaker")
        if "'Headphone',0" in amixer_output.decode("utf-8"):
            channel_list.append("Headphone")

        return channel_list

    def init_dbus(self):
        try:
            return blockifydbus.BlockifyDBus()
        except Exception as e:
            log.error("Cannot connect to DBus. Exiting.\n ({}).".format(e))
            sys.exit()

    def refresh_spotify_process_state(self):
        """Check if Spotify is running periodically. If it's not, suspend blockify."""
        previous_suspend_state = self.suspend_blockify
        if not self.check_for_spotify_process():
            self.suspend_blockify = True
        else:
            self.suspend_blockify = False

        if previous_suspend_state is not self.suspend_blockify:
            if not self.suspend_blockify:
                self.dbus.connect_to_spotify_dbus(None)
                self.player.try_resume_spotify_playback(True)
                log.warn("Spotify was restarted! Connecting now.")
            else:
                log.warn("Spotify was closed!")

        return True

    def start(self):
        self.bind_signals()
        # Force unmute to properly initialize unmuted state

        self.toggle_mute(2)

        GObject.timeout_add(self.spotify_refresh_interval, self.refresh_spotify_process_state)
        GObject.timeout_add(self.update_interval, self.update)
        if self.autoplay:
            # Delay autoplayback until self.spotify_is_playing was called at least once.
            GObject.timeout_add(self.update_interval + 100, self.start_autoplay)

        log.info("Blockify started.")

        Gtk.main()

    def start_autoplay(self):
        if self.autoplay:
            log.debug("Autoplay is activated.")
            log.info("Starting Spotify autoplayback.")
            self.dbus.play()
        return False

    def adjust_interlude(self):
        if self.use_interlude_music:
            self.player.toggle_music()

    def spotify_is_playing(self):
        return self.song_status == "Playing"

    def update(self):
        "Main update routine, looped every self.update_interval milliseconds."
        if not self.suspend_blockify:
            # Determine if a commercial is running and act accordingly.
            self.found = self.find_ad()

            self.adjust_interlude()

        # Always return True to keep looping this method.
        return True

    def find_ad(self):
        "Main loop. Checks for ads and mutes accordingly."
        self.previous_song = self.current_song
        self.current_song = self.get_current_song()
        # self.song_status = self.dbus.get_song_status()

        # Manual control is enabled so we exit here.
        if not self.automute:
            return False

        if self.autodetect and self.current_song and self.current_song_is_ad():
            if self.use_interlude_music and not self.player.temp_disable:
                self.player.temp_disable = True
                GObject.timeout_add(self.player.playback_delay, self.player.play_with_delay)
            self.ad_found()
            return True

        # Check if the blockfile has changed.
        try:
            current_timestamp = self.blocklist.get_timestamp()
        except OSError as e:
            log.debug("Failed reading blocklist timestamp: {}. Recovering.".format(e))
            self.blocklist.__init__()
            current_timestamp = self.blocklist.timestamp
        if self.blocklist.timestamp != current_timestamp:
            log.info("Blockfile changed. Reloading.")
            self.blocklist.__init__()

        if self.blocklist.find(self.current_song):
            self.ad_found()
            return True

        # Unmute with a certain delay to avoid the last second
        # of commercial you sometimes hear because it's unmuted too early.
        GObject.timeout_add(self.unmute_delay, self.unmute_with_delay)

        return False

    def ad_found(self):
        # log.debug("Ad found: {0}".format(self.current_song))
        self.toggle_mute(1)

    def current_song_is_ad(self):
        "Compares the wnck song info to dbus song info."
        try:
            song_artist = self.dbus.get_song_artist()
            song_title = self.dbus.get_song_title()
            try:
                song_artist = song_artist.decode("utf-8")
                song_title = song_title.decode("utf-8")
            except AttributeError as e:
                log.debug("AttributeError during ad detection: {}".format(e))
            dbus_song = song_artist + self.song_delimiter + song_title
            is_ad = self.current_song != dbus_song
            return is_ad
        except TypeError as e:
            # Spotify has technically stopped playing and sending dbus
            # metadata so we get NoneType-errors.
            # However, it might still play one last ad so we assume that
            # is the case here.
            log.debug("TypeError during ad detection: {}".format(e))
            return True

    def unmute_with_delay(self):
        if not self.found:
            self.toggle_mute()
        return False

    def find_spotify_window_wmctrl(self):
        spotify_window = []
        try:
            pipe = subprocess.Popen(['wmctrl', '-lx'], stdout=subprocess.PIPE).stdout
            window_list = pipe.read().decode("utf-8").split("\n")
            for window in window_list:
                if window.find("spotify.Spotify") >= 0:
                    # current_song = " ".join(window.split()[5:])
                    spotify_window.append(window)
                    break

        except OSError:
            log.error("Please install wmctrl first! Exiting.")
            sys.exit(1)

        return spotify_window

    def find_spotify_window(self):
        "Libwnck list of currently open windows."
        Gtk.init([])
        # Get the current screen.
        screen = Wnck.Screen.get_default()

        # recommended per Wnck documentation
        screen.force_update()

        # Object list of windows in screen.
        windows = screen.get_windows()

        # Return the Spotify window or an empty list.
        return [win.get_icon_name() for win in windows\
                if len(windows) and "Spotify" in win.get_application().get_name()]

    def get_current_song(self):
        "Checks if a Spotify window exists and returns the current songname."
        song = ""
        spotify_window = self.find_spotify_window_wmctrl()
        if spotify_window:
            try:
                song = " ".join(map(str, spotify_window[0].split()[4:]))
            except Exception as e:
                log.debug("Could not match spotify pid to sink pid: %s".format(e), exc_info=1)

        return song

    def block_current(self):
        if self.current_song:
            self.blocklist.append(self.current_song)

    def unblock_current(self):
        if self.current_song:
            if self.use_interlude_music:
                self.player.pause()
            song = self.blocklist.find(self.current_song)
            if song:
                self.blocklist.remove(song)
            else:
                log.error("Not found in blocklist or block pattern too short.")

    def toggle_mute(self, mode=0):
        # 0 = automatic, 1 = force mute, 2 = force unmute
        self.mutemethod(mode)

    def is_muted(self):
        for channel in self.channels:
            output = subprocess.check_output(["amixer", "get", channel])
            if "[off]" in output:
                return True
        return False

    def get_state(self, mode):
        muted = self.is_muted()
        self.is_fully_muted = muted

        state = None

        if muted and (mode == 2 or not self.current_song):
            state = "unmute"
        elif muted and mode == 0:
            state = "unmute"
            log.info("Unmuting.")
        elif not muted and mode == 1:
            state = "mute"
            log.info("Muting {}.".format(self.current_song))

        return state

    def alsa_mute(self, mode):
        "Mute method for systems without Pulseaudio. Mutes sound system-wide."
        state = self.get_state(mode)
        if not state:
            return

        for channel in self.channels:
            subprocess.Popen(["amixer", "-q", "set", channel, state])

    def pulse_mute(self, mode):
        "Used if pulseaudio is installed but no sinks are found. System-wide."
        state = self.get_state(mode)
        if not state:
            return

        for channel in self.channels:
            subprocess.Popen(["amixer", "-qD", "pulse", "set", channel, state])

    def extract_pulse_sink_status(self, pacmd_out):
        sink_status = ("", "")
        # Match muted_value and application.process.id values.
        pattern = re.compile(r"(?: index|muted|application\.process\.id).*?(\w+)")
        # Put valid spotify PIDs in a list
        output = pacmd_out.decode("utf-8")

        spotify_sink_list = [" index: " + i for i in output.split("index: ") if "spotify" in i]

        if len(spotify_sink_list) and self.spotify_pids:
            sink_infos = [pattern.findall(sink) for sink in spotify_sink_list]
            # Every third element per sublist is a key, the value is the preceding
            # two elements in the form of a tuple - {pid : (index, muted_value)}
            idxd = {sink_status[2]: (sink_status[0], sink_status[1]) for sink_status in sink_infos if len(sink_status) == 3}

            pid = [k for k in idxd.keys() if k in self.spotify_pids][0]
            sink_status = idxd[pid]  # tuple of 2 elements: (index, muted_value)

        return sink_status

    def pulsesink_mute(self, mode):
        "Finds spotify's audio sink and toggles its mute state."
        try:
            pacmd_out = subprocess.check_output(["pacmd", "list-sink-inputs"])
        except subprocess.CalledProcessError:
            log.error("Spotify sink not found. Is Pulse running? Resorting to pulse amixer as mute method.")
            self.mutemethod = self.pulse_mute  # Fall back to amixer mute.
            self.use_interlude_music = False
            return

        index, muted_value = self.extract_pulse_sink_status(pacmd_out)

        self.is_sink_muted = False if muted_value == self.pulse_unmuted_value else True

        if index:
            if self.is_sink_muted and (mode == 2 or not self.current_song):
                log.info("Forcing unmute.")
                subprocess.call(["pacmd", "set-sink-input-mute", index, "0"])
            elif not self.is_sink_muted and mode == 1:
                log.info("Muting {}.".format(self.current_song))
                subprocess.call(["pacmd", "set-sink-input-mute", index, "1"])
            elif self.is_sink_muted and not mode:
                log.info("Unmuting.")
                subprocess.call(["pacmd", "set-sink-input-mute", index, "0"])

    def prev(self):
        self.dbus.prev()
        self.player.try_resume_spotify_playback()

    def next(self):
        self.dbus.next()
        self.player.try_resume_spotify_playback()

    def signal_stop_received(self, sig, hdl):
        log.debug("{} received. Exiting safely.".format(sig))
        self.stop()

    def signal_block_received(self, sig, hdl):
        log.debug("Signal {} received. Blocking current song.".format(sig))
        self.block_current()

    def signal_unblock_received(self, sig, hdl):
        log.debug("Signal {} received. Unblocking current song.".format(sig))
        self.unblock_current()

    def signal_prev_received(self, sig, hdl):
        log.debug("Signal {} received. Playing previous interlude.".format(sig))
        self.prev()

    def signal_next_received(self, sig, hdl):
        log.debug("Signal {} received. Playing next song.".format(sig))
        self.next()

    def signal_playpause_received(self, sig, hdl):
        log.debug("Signal {} received. Toggling play state.".format(sig))
        self.dbus.playpause()

    def signal_toggle_block_received(self, sig, hdl):
        log.debug("Signal {} received. Toggling blocked state.".format(sig))
        self.toggle_block()

    def signal_prev_interlude_received(self, sig, hdl):
        log.debug("Signal {} received. Playing previous interlude.".format(sig))
        self.player.prev()

    def signal_next_interlude_received(self, sig, hdl):
        log.debug("Signal {} received. Playing next interlude.".format(sig))
        self.player.next()

    def signal_playpause_interlude_received(self, sig, hdl):
        log.debug("Signal {} received. Toggling interlude play state.".format(sig))
        self.player.playpause()

    def signal_toggle_autoresume_received(self, sig, hdl):
        log.debug("Signal {} received. Toggling autoresume.".format(sig))
        self.player.toggle_autoresume()

    def bind_signals(self):
        "Catch signals because it seems like a great idea, right? ... Right?"
        signal.signal(signal.SIGINT, self.signal_stop_received)  # 9
        signal.signal(signal.SIGTERM, self.signal_stop_received)  # 15

        signal.signal(signal.SIGUSR1, self.signal_block_received)  # 10
        signal.signal(signal.SIGUSR2, self.signal_unblock_received)  # 12

        signal.signal(signal.SIGRTMIN, self.signal_prev_received)  # 34
        signal.signal(signal.SIGRTMIN + 1, self.signal_next_received)  # 35
        signal.signal(signal.SIGRTMIN + 2, self.signal_playpause_received)  # 35
        signal.signal(signal.SIGRTMIN + 3, self.signal_toggle_block_received)  # 37

        signal.signal(signal.SIGRTMIN + 10, self.signal_prev_interlude_received)  # 44
        signal.signal(signal.SIGRTMIN + 11, self.signal_next_interlude_received)  # 45
        signal.signal(signal.SIGRTMIN + 12, self.signal_playpause_interlude_received)  # 46
        signal.signal(signal.SIGRTMIN + 13, self.signal_toggle_autoresume_received)  # 47

    def stop(self):
        log.info("Exiting safely. Bye.")
        # Stop the interlude player.
        if self.use_interlude_music:
            self.use_interlude_music = False
            self.player.stop()
        # Save the list only if it changed during runtime.
        if self.blocklist != self.orglist:
            self.blocklist.save()
        # Unmute before exiting.
        self.toggle_mute(2)
        sys.exit()

    def toggle_block(self):
        "Block/unblock the current song."
        if self.found:
            self.unblock_current()
        else:
            self.block_current()
            if self.use_interlude_music:
                self.player.manual_control = False

    @property
    def automute(self):
        return self._automute

    @automute.setter
    def automute(self, boolean):
        log.debug("Automute: {}.".format(boolean))
        self._automute = boolean

    @property
    def autodetect(self):
        return self._autodetect

    @autodetect.setter
    def autodetect(self, boolean):
        log.debug("Autodetect: {}.".format(boolean))
        self._autodetect = boolean


def initialize(doc=__doc__):
    util.initialize(doc)

    blockify = Blockify(blocklist.Blocklist())

    return blockify


def main():
    "Entry point for the CLI-version of Blockify."
    blockify = initialize()
    blockify.start()


if __name__ == "__main__":
    main()
